providers/implementations/keymgmt/mlx_kmgmt.c

```diff
@@ -0,0 +1,764 @@
+        ppkey = &key->mkey;
+        off = slot * xbytes;
+        len = mbytes;
+    } else {
+        alg = key->xinfo->algorithm_name;
+        group = (char *) key->xinfo->group_name;
+        ppkey = &key->xkey;
+        off = (1 - ml_kem_slot) * mbytes;
+        len = xbytes;
+    }
+    val = (void *)(in + off);
+
+    if ((ctx = EVP_PKEY_CTX_new_from_name(libctx, alg, propq)) == NULL
+        || !EVP_PKEY_fromdata_init(ctx))
+        goto err;
+    parr[0] = OSSL_PARAM_construct_octet_string(pname, val, len);
+    if (group != NULL)
+        parr[1] = OSSL_PARAM_construct_utf8_string(OSSL_PKEY_PARAM_GROUP_NAME,
+                                                   group, 0);
+    if (EVP_PKEY_fromdata(ctx, ppkey, selection, parr))
+        ret = 1;
+
+  err:
+    EVP_PKEY_CTX_free(ctx);
+    return ret;
+}
+
+static int
+load_keys(MLX_KEY *key,
+          const uint8_t *pubenc, size_t publen,
+          const uint8_t *prvenc, size_t prvlen)
+{
+    int slot;
+    int selection;
+
+    selection = OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS;
+    selection |= OSSL_KEYMGMT_SELECT_KEYPAIR;
+
+    for (slot = 0; slot < 2; ++slot) {
+        if (prvlen) {
+            /* Ignore public keys when private provided */
+            if (!load_slot(key->libctx, key->propq, OSSL_PKEY_PARAM_PRIV_KEY,
+                           selection, key, slot, prvenc,
+                           key->minfo->prvkey_bytes, key->xinfo->prvkey_bytes))
+                goto err;
+        } else if (publen) {
+            /* Absent private key data, import public keys */
+            if (!load_slot(key->libctx, key->propq, OSSL_PKEY_PARAM_PUB_KEY,
+                           selection, key, slot, pubenc,
+                           key->minfo->pubkey_bytes, key->xinfo->pubkey_bytes))
+                goto err;
+        }
+    }
+    key->state = prvlen ? MLX_HAVE_PRVKEY : MLX_HAVE_PUBKEY;
+    return 1;
+
+  err:
+    EVP_PKEY_free(key->mkey);
+    EVP_PKEY_free(key->xkey);
+    key->xkey = key->mkey = NULL;
+    key->state = MLX_HAVE_NOKEYS;
+    return 0;
+}
+
+static int mlx_kem_key_fromdata(MLX_KEY *key,
+                               const OSSL_PARAM params[],
+                               int include_private)
+{
+    const OSSL_PARAM *param_prv_key = NULL, *param_pub_key;
+    const void *pubenc = NULL, *prvenc = NULL;
+    size_t pubkey_bytes, prvkey_bytes;
+    size_t publen = 0, prvlen = 0;
+
+    /* Invalid attempt to mutate a key, what is the right error to report? */
+    if (key == NULL || mlx_kem_have_pubkey(key))
+        return 0;
+    pubkey_bytes = key->minfo->pubkey_bytes + key->xinfo->pubkey_bytes;
+    prvkey_bytes = key->minfo->prvkey_bytes + key->xinfo->prvkey_bytes;
+
+    /* What does the caller want to set? */
+    param_pub_key = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PUB_KEY);
+    if (param_pub_key != NULL &&
+        OSSL_PARAM_get_octet_string_ptr(param_pub_key, &pubenc, &publen) != 1)
+        return 0;
```

Comment:
ERR_raise_data to indicate *specifically* what is missing?
Alternatively allow the catching of this error further down ... where the lengths are checked?

Reply:
This "error" is a parameter *decoding* error, with the PARAM layer responsible for augmenting the error stack (wrong type is the only reason that comes to mind).